### ISSUE TITLE
Better handling of CMS pages in a multishop environment

### ISSUE THREAD

**Issue by lmeyer1** [PrestaShop/PrestaShop#25478]
#### Is your feature request related to a problem?

There is a problem with CMS in multishop context. In the table `cms_lang` there is one version of the text for each shop. This is OK, but gives rise to following problems: 
- When in *All shops* context, a change affects all versions. When in a *specific shop*, only the version of that shop is affected by modifications. This is a dangerous behavior, because one can easily make one of these two mistakes:
    - If the page is customized in all shops, one can destroy it when in *All shops*
    - If the page is the same everywhere, one can introduce easily inconsistencies by changing the content in only one shop.
- Of much less importance: I f the content is the same everywhere, or if the page is used in only one shop, there is redundancy, because the information is stored for every shop and every language regardless of if it is used.

#### Describe the solution you'd like

In the list of pages, indicate if the content is identical *Identical everywhere* in all shops or if it is different *Different from shop to shop*. This flag relies directly on the content stored in `cms_lang`. The flag must reflect only the content of active shop and languages.

If the content is *Identical everywhere*:
- In *All shops* context:
   - A user can only edit the page if he has access to all the shops where the page is published
   - *Associated shops* can be freely changed (respecting the access permissions of the user).
- In specific shop context:
   - Instead of the fields allowing to edit the page, display a warning recommending to edit in *All shops* context. Show a link that allows to edit the page in specific shop contexts. Once modified in the specific shop, the page will have the flag *Different from shop to shop*
- In shop group context: ask the user to choose *All shops*

If the content is *Different from shop to shop*:
- In specific shop context:
   - It is sufficient for the user to have access to his own shop to edit the page content.
   - The user can change *Associated shops* only for the shops he has permissions for. When checking a new shop, the content of the current shop is copied to the new one.
- In *All shops* context:
   - Instead of the fields allowing to edit the page, display a warning recommending to edit in specific shop context. If the user has sufficient permissions, show a link that allows to copy the page from the first shop context to all other shops and thus to override any shop specific versions (it would even be better to be able to select the shop the data is copied from). 
- In shop group context: ask the user to choose specific shop.

#### Alternatives you've considered

If there are many stores with many people intervening, errors might happen without noticing. The above rules make sure, no one makes unintentionally destructive changes. 

#### Additional context

Feature request discussed on slack.
I hope the proposed changes are general enough to not introduce unwanted limitations in any use case.

**Comment by hibatallahAouadni** [comment]
Hello @lmeyer1 


Thank you for your suggestion and the well detailed report.
The Product Team will take it into consideration for future developments. 
Please be aware that there is no guarantee that this feature will be developed anytime soon. If this is important to you, we strongly suggest you consider [hiring a specialist](https://www.prestashop.com/en/experts) to help you create it. 

If you are willing to implement this feature into PrestaShop as a [Pull Request](https://devdocs.prestashop.com/1.7/contribute/contribute-pull-requests/), the teams will be happy to help you by ensuring that it is appropriate and coherent with the rest of the project. If that is the case, please let us know! 

Thank you

**Comment by marionf** [comment]
Hello @lmeyer1  and thank you for your comments

> - When in *All shops* context, a change affects all versions. When in a *specific shop*, only the version of that shop is affected by modifications. This is a dangerous behavior, because one can easily make one of these two mistakes.

In fact, it's not when you are in "All shops" context, it rather depends on the store(s) selected in the shop association block.
You can be in "all shops" context but then deselect shop(s) in the shop association block.

>  - If the page is customized in all shops, one can destroy it when in *All shops*

That's the goal of the all shops context: make a change that will affect all stores.

> - If the page is the same everywhere, one can introduce easily inconsistencies by changing the content in only one shop.

Unless if there is a bug that I am not aware of, this is not possible because if the page is the same in all shops and if you edit it only in the context of 1 shop, all the shops are checked in the shop association block, so the content remains the same for all shops.
And if you check only one shop in the shop association block, then the page is not associated anymore with the shop(s) unselected.

---

When a page is edited, the modifications made are applied to all the stores with which the page is associated. This way, the page is always the same for all the shops with which it is associated: we don't let the user the possibility of having different content for a page associated with several shops.

Example: 

Let's suppose I have 2 shops: "Shop1" and "Shop2".

I create a page called "Terms and conditions" and I associate it with all stores through the shop association block.

In the listing in all shops context I have:

"Terms and conditions" page which is associated with "Shop1" and "Shop2".

Finally, I realize that in "Shop2" the terms and conditions are not the same as in Shop1.

I have to edit the "Terms and conditions" page and uncheck the "shop2" box, then I have to create a new "Terms and conditions" page with different content, associated only with "Shop2".

Now, in the listing in all shops context I have:

- A page "Terms and conditions" which is associated with "Shop1"
- Another page "Terms and conditions" which is associated with "Shop2"

If tomorrow, I want to have the same terms and conditions for both shops, I will have to delete one of the two pages and edit the other one to associate it with the other shop.

---

The behavior I described above is the same for several back-office pages: add/edit CMS page, category, brand, supplier, languages, currencies, zones, contact, store ... and more
We can't change the behavior only for the CMS page and not for the other BO pages, this would create inconsistency between the different pages of the back office and risk confusing users.

Some users have told us that they made configuration errors with the wrong context. This should be resolved in 1.7.8 with [the new multistore header](https://github.com/PrestaShop/PrestaShop/pull/23292) and the colors that can be associated with the different shops.

**Comment by lmeyer1** [comment]
@marionf 
Thanks for your detailed explanations.
I don't agree in every point though. Let me explain.

> When a page is edited, the modifications made are applied to all the stores with which the page is associated.

This is only the case when editing in *All shop* context. This is not true when editing in a specific shop. In that case, only the text in that specific shop is modified.

> Example:
Let's suppose I have 2 shops: "Shop1" and "Shop2".
(...)

Yes this is how it worked in PS 1.6.0, and it is still possible to use CMS that way today.

At some place in development history (as far as I remember since version 1.6.1.1), the content of a page was saved in as many versions as there are shops (before there was only one version). This led me to think that the intended use for a page like "Terms and conditions" that is different in "Shop 1" and "Shop 2" is to leave the content under one page, but customize its content in the two shops. 

My opinion about this is : both ways to use customize CMS pages should work:
- Create a new page for each customization of the content (as you explained very well)
- Do not create a new page, but customize the content by shop.

In fact, if the second way to use CMS pages were **NOT** intended, then the change in database structure implemented around 1.6.1.1. was clearly a mistake.

> Some users have told us that they made configuration errors with the wrong context. This should be resolved in 1.7.8

Sure, the new multishop display in BO will reduce the chance to make mistakes, but my PR will add more safeguards. It also should never interfere when using CMS the way you explained above.

For me this feature of **having one version of a page for each shop** is important. It allows to manage shop configuration:
- The link to accept the terms and conditions is the same in every shop.
- The link lists in the footer are the same.

This is not an issue, if you are working with 2 shops, but it becomes crucial when working with dozens of shops. We are at 25 shops on one installation today.

**Comment by marionf** [comment]
> When a page is edited, the modifications made are applied to all the stores with which the page is associated.

> This is only the case when editing in All shop context. This is not true when editing in a specific shop. In that case, only the text in that specific shop is modified.

You are right, it seems to be a bug. According to the [current specifications](https://github.com/PrestaShop/prestashop-specs/blob/master/content/1.7/back-office/multistoregeneralspecs.md#actions-in-the-listings-grids), the content should stay the same for every shop checked in the shop association block.

> My opinion about this is : both ways to use customize CMS pages should work:
> - Create a new page for each customization of the content (as you explained very well)
> - Do not create a new page, but customize the content by shop.

The second option to not create a new page but let the user customize the content by shop could be a great improvement in the multistore management. 
But it's a big change that we can't implement without doing some user testing first: we have to make sure that this new way of managing multistore will be well understood by everyone. Also, if we decide to implement it, we will have to do it on all the pages of the back office to keep consistency.

Unfortunately, for now, the multistore is not anymore a priority as we already worked on it for 1.7.8.
By the way, if you want to contribute to improving the multistore feature, you can help to finish the [implementation of drop-downs and checkboxes](https://github.com/PrestaShop/PrestaShop/issues/25058) on the migrated pages.

The product team is now focused on the [next major version](https://github.com/PrestaShop/PrestaShop/issues/24111).
But I let this issue open if one day we resume the research work on the multistore to improve it.

Thank you very much for sharing your need and your point of view 👍
